packages/clerk-js/src/ui/common/withRedirectToHome.tsx

```diff
@@ -1,35 +1,67 @@
 import { SignInProps, SignUpProps } from '@clerk/types';
-import React, { useContext } from 'react';
+import React from 'react';
 
-import { useEnvironment } from '../contexts';
-import { CoreSessionContext } from '../contexts/CoreSessionContext';
+import { canShowOrganizationProfile, canShowSignIn, canShowSignUp, canShowUserProfile } from '../../utils';
+import { useCoreClerk, useEnvironment } from '../contexts';
 import { useNavigate } from '../hooks';
 
 export function withRedirectToHome<P extends SignInProps | SignUpProps | { continueExisting?: boolean }>(
   Component: React.ComponentType<P>,
-  displayName?: string,
+  when: 'signIn' | 'signUp' | 'userProfile' | 'organizationProfile',
 ): (props: P) => null | JSX.Element {
-  displayName = displayName || Component.displayName || Component.name || 'Component';
+  const displayName = Component.displayName || Component.name || 'Component';
   Component.displayName = displayName;
 
   const HOC = (props: P) => {
     const { navigate } = useNavigate();
-    const { authConfig, displayConfig } = useEnvironment();
-    const { singleSessionMode } = authConfig;
-    const ctx = useContext(CoreSessionContext);
-    const session = ctx?.value;
+    const clerk = useCoreClerk();
+    const environment = useEnvironment();
+
+    let shouldRedirect = false;
+    let warning = '';
+    switch (when) {
```

Comment:
⛔ We can make this method more dry by making each Component aware of whether it can be rendered. That is, let's introduce a `Component.shouldRender` static method that accepts (clerk, and environment) and calls that method here instead of doing a switch statement. That way, you will also avoid the second string parameter passed `withRedirectToHome` representing the component name.

Reply:
I'd suggest something along the lines of what Sokratis said above, as this `switch` statement here can be considered a code smell: when a new component is added, we'll have to find all similar `switch` statements and update them.

A different, more "FP-oriented" approach would be to pass a condition as an argument to the `withRedirectToHome` HOC, creating a more specific HOC as a result. eg;

```
const withRedirectToHomeUserGuard = (...args) => withRedirectToHome(condition, ...args)
const withRedirectToHomeOrganizationGuard = (...args) => withRedirectToHome(condition, ...args)
```

The names are terrible, I know, the rest is up to you :)

Reply:
@SokratisVidros We would have to create a lot of `shouldRender` functions, since we use the HOC in most of the routes for `SignIn`, but not all. It does not wrap the `SignIn` but rather some of its route components.

Reply:
Also, is there a reason we can't make this check in the `mount` methods? What's preventing us from doing that?

Reply:
I did what @nikosdouvlis proposed since it is a simpler approach for now (because we use the HOC many times) and I think it scales well. Please take a look.